[Pyro] 미션3: 체스판 초기화

ghojeong opened the pull request:
Main 클래스를 실행하면, 아래와 같이 결과가 나옵니다.

<img width="320" alt="스크린샷 2021-02-17 오후 5 35 50" src="https://user-images.githubusercontent.com/38595650/108180339-fca2a180-7149-11eb-8427-0271fbba5b7c.png">

미션3를 진행하며 제일 고심했던 것은 Board 를 어떻게 구현을 할 것인지였습니다.

처음에는 아무 생각없이 Piece 멤버를 Position 클래스가 갖도록 하고,
Position의 리스트를 Row 로 만들고,
Row 들의 리스트를 Board 로 만들어야겠다!! 
그리고 Piece 를 옮길 때는 Board 안의 Position 내부의 Piece 멤버를 수정해야겠다! 
라고 생각을 했습니다.

막상 Pseudo 코드를 작성하니, Piece 를 옮기기 위해서,
Board 부터 Position 까지 클래스를 타고 내려가는 과정이 매우 번거로웠습니다.

어떻게 더 간략화 할 방법이 없을 까 고심하다가,
Position 와 Piece 를 서로 분리시키고,
이 둘을 Board 에서 Map 자료 구조를 통해 연결시키자는 결론을 내렸습니다.

이렇게 하면 해리포터와 마법사의 돌에 나오는 마법체스를, 론이 플레이할 때처럼
"d5 에 있는 말을 d2 로 옮겨라!" 같은 구두를 통한 플레이가 가능합니다.

Commit:
[refactor]: 테스트 코드의 접근 제한자 삭제

Commit:
[test] Pawn 색에 따른 출력 문자 부여 테스트

- 흰색이면 소문자, 까만색이면 대문자여야 한다.
Lucas 의 테스트 코드를 그대로 사용할까 고민하다가,
Open Closed Principle 을 위해서 기존 테스트 코드를 수정하지 않고,
색에 따른 출력 문자 부여 테스트를 추가할 수 있도록 해보았다.

Commit:
[feat] Pawn 색에 따른 출력 문자 부여 구현

조건문 없이, representationStrategy 를 통해
출력 문자가 다르게 표현되도록 했다.

Commit:
[test] Add PositionTest, BoardViewTest

보드를 출력하는 BoardView 와
각 Piece 의 위치정보를 나타낼 Position의 테스트를 추가했다.
Board 는 Position 과 Piece 의 관계를 나타내는 자료구조로 구현해 나가야한다.

Commit:
[feat] Board 와 BoardView 구현

Board 에서 Map 자료구조를 사용한 이유는 나중에 구현할
Piece 를 움직이는 로직을 위해서이다.
d5 에 있는 말을 f2 로 옮겨라 와 같은 커맨드를 실행할려면,
해당 Position 에 있는 Piece 가 무엇이고, 이동하는 곳에 어떤 Piece 가 있는지 알아야만 한다.

Commit:
[feat] main 메소드 구현

- start 라는 명령을 입력하면 체스 게임을 시작하고, 체스판의 현재 상태를 콘솔 화면에 출력한다.
- end 라는 명령을 입력하면 체스 게임을 종료한다.

Commit:
[chore] BoardTest::initialize 추가

Lucas 에 나와있는 의미 불명의 테스트 추가.
getWhitePawnsResult 와 getBlackPawnsResult 둘다 불필요해보이므로,
나중에 필히 삭제해야함.
이 커밋은 가능한 revert 되었으면 좋겠음

Commit:
[test] DisplayName 보완

Commit:
[refactor] Piece 의 representation 멤버를 삭제

getIcon 이라는 추상 함수로, 하위 클래스들이 직접 표현하도록 함
getIcon 이 public 이 아닌 이유는, pieces 패키지 바깥에서 사용되기를 원하지 않기 때문이다.

ghojeong commented on src/main/java/chess/board/Board.java:
```diff
@@ -0,0 +1,71 @@
+public class Board {
+    private final Map<Position, Piece> squares = new LinkedHashMap<>();
+
+    public void initialize() {
+        addRoyalRow(8, Color.BLACK);
+        addPawnRow(7, Color.BLACK);
+        for (int i = 6; i >= 3; i--) {
+            addEmptyRow(i);
+        }
+        addPawnRow(2, Color.WHITE);
+        addRoyalRow(1, Color.WHITE);
+    }
```
initialize 와 관련된 로직들은 나중에 BoardFactory 클래스로 분리할 예정입니다.
Board 클래스 에는 게임을 진행하며 Piece 를 움직일 때 발생하는 게임판위의 이벤트들을 처리하는 로직을 담으려 합니다.

ghojeong commented:
8 과 같은 상수를 별도의 static final  값으로 정의해야할까 고민하다가,
어차피 체스판은 8 x 8 보드 사이즈에서 규칙과 보드가 바뀔일이 존재하지 않으므로,
8 과 같은 값을 하드코딩을 하더라도 크게 문제 없지 않을까 라고 생각하게 되었습니다.

Commit:
[refactor] 불필요한 String + 연산 제거

ghojeong commented on src/main/java/chess/view/BoardView.java:
```diff
@@ -0,0 +1,54 @@
+        sb.append(StringUtils.NEWLINE);
+    }
+
+    private String getRepresentation(List list) {
+        StringBuilder sb = new StringBuilder();
+        Iterator itr = list.iterator();
```
Polymorphism 을 위해, 일부러 제네릭 타입을 명시하지 않았습니다.

honux77 commented on src/main/java/chess/Main.java:
```diff
@@ -0,0 +1,36 @@
+        System.out.println("체스 게임을 시작합니다.");
+        System.out.println("게임 시작은 start, 종료는 end 명령을 입력하세요.");
+
+        Scanner scanner = new Scanner(System.in);
+        String command = getCommand(scanner);
+        if (!command.equals("start")) {
+            scanner.close();
+            return;
+        }
+
+        Board board = new Board();
+        board.initialize();
+        BoardView boardView = new BoardView(board);
+
+        while (!command.equals("end")) {
+            System.out.println(boardView.getBoardRepresentation());
+            command = getCommand(scanner);
+        }
+
+        scanner.close();
+    }
```
요구사항에는 없지만 없는 명령을 받으면 종료보다는 "잘못 입력하셨습니다" 하고 다시 되는게 조금 더 낫지 않을까요?

honux77 replied:
그리고 Main보다는 ChessGame 같은 클래스를 만들고 main에서는 가급적 로직을 제거하고 메소드들을 호출하는 것도 좋을 것 같아요.

ghojeong replied:
> "잘못 입력하셨습니다" 하고 다시 되는게 조금 더 낫지 않을까요?

게임 플레이의 디테일이군요. 해당 기능 추가로 구현하겠습니다.

ghojeong replied:
>  ChessGame 같은 클래스를 만들고 main에서는 가급적 로직을 제거하고 메소드들을 호출하는 것도 좋을 것 같아요.

아! 확실히 그렇네요. ChessGame 클래스 별도로 만들겠습니다.

honux77 commented on src/main/java/chess/board/Board.java:
```diff
@@ -0,0 +1,71 @@
+import static java.util.stream.Collectors.toList;
+
+public class Board {
+    private final Map<Position, Piece> squares = new LinkedHashMap<>();
```
오 요거 괜찮네요. 저는 Map<Rank, List<Piece>> 로 했는데 파이로 처럼 하는게 더 좋을 것 같습니다.

janeljs replied:
호오옹...저도 따라해봐야 겠어요. 역시 파이로👍👍

honux77 commented on src/main/java/chess/board/Board.java:
```diff
@@ -0,0 +1,71 @@
+    public void initialize() {
+        addRoyalRow(8, Color.BLACK);
+        addPawnRow(7, Color.BLACK);
+        for (int i = 6; i >= 3; i--) {
+            addEmptyRow(i);
+        }
```
저는 그냥 이런거 하드 코딩을 좋아해요. 세번 복붙이 더 깔끔한 느낌?

honux77 replied:
4번이구나 ㅋㅋㅋ. 
앗 근데 왜 거꾸로 했나 생각했더니 출력 방향 때문이군요. 
흠.. 그것보다는 순서대로 추가해주고 출력에서 반대로 뒤집어 보는건 어때요??

Commit:
[refactor] Piece 에서 representation 멤버 부활시킴

generateRepresentation 메소드가 매번 호출되는게 낭비라고 생각되어서,
repersentation 멤버를 초기화시 저장하도록 하도록 바꾸었습니다.

honux77 commented on src/main/java/chess/board/Board.java:
```diff
@@ -0,0 +1,71 @@
+
+    private void addEmptyRow(int rowId) {
+        for (int i = 'a'; i <= 'h'; i++) {
+            char columnId = (char) i;
```
for (char i = 'a';...)

honux77 replied:
char는 아마 평범한 부호있는 정수일껄요?

Dae-Hwa replied:
> for (char i = 'a';...)

와 이거 좋네요!

honux77 commented on src/main/java/chess/board/Board.java:
```diff
@@ -0,0 +1,71 @@
+    private void addPawnRow(int rowId, Color color) {
+        for (int i = 'a'; i <= 'h'; i++) {
+            char columnId = (char) i;
+            squares.put(new Position(columnId, rowId), new Pawn(color));
```
Postion.create(col, row) 로 해 보는 건 어때요?

ghojeong replied:
static factory 메소드를 통해 생성할 수 있도록 하겠습니다.

Commit:
[refactor] 루프를 하드 코딩으로 변환; for (int i = 6; i >= 3; i--)

honux77 commented on src/main/java/chess/board/Board.java:
```diff
@@ -0,0 +1,71 @@
+
+    private void addRoyalRow(int rowId, Color color) {
+        addEmptyRow(rowId);
+    }
```
요거는 나중에 업데이트 하는 거죠? 잊지 않게 
//TODO: implement

정도 주석 남겨 놓는 것도 좋겠어요. IDE 색깔도 달라집니다.

honux77 commented on src/main/java/chess/board/Board.java:
```diff
@@ -0,0 +1,71 @@
+                .stream()
+                .map(Entry::getValue)
+                .collect(toList());
+    }
```
뭔가 좀 복잡한데요? 더 좋은 방법 없을까요? 
그리고 public 메소드가 너무 많은데 왠지 제가 보기엔 public 일 필요가 없는 것 같아요.

honux77 replied:
제가 보기엔 내부적으로 Map만 가지고 있는데 List를 매번 만드니까 생기는 문제 같은데요. 
Map과 리스트를 동시에 유지하고 있으면 코드의 복잡성을 줄이고 성능을 개선할 수 있지 않을까요?

honux77 replied:
아니면 return new ArrayList<>(squares.values());

honux77 replied:
다시 생각해 보니 그냥 지금 이대로가 좋은 것 같아요. 추가적인 자료구조를 복잡하게 만드는 것보다 지금 코드가 더 낫네요. 여튼 저 부분은 고쳐야겠죠?

ghojeong replied:
getPieces 가 존재하지 않도록 해당 부분을 싸그리 뜯어고쳤습니다.

Commit:
[refactor] 패키지 구조 변경; chess.Main -> chess.controller.ChessGame

board 와 peieces 패키지를 domain 패키지 안으로 넣음
미약하게나마 view, domain, controller 분리를 시도함

honux77 commented on src/main/java/chess/board/Board.java:
```diff
@@ -0,0 +1,71 @@
+        getPieces().subList(48, 56)
+                .forEach(sb::append);
+        return sb.toString();
+    }
```
48, 56 같은 하드 코딩은 지양하는 편이 좋은데요. 차라리 final int start = 48 이런식으로 지역변수라도 쓰는게 좋을까 싶어요.

ghojeong replied:
getWhitePawnsReulst 함수가 매우 마음에 들지 않아서, 관련된 public 함수들을 전부 삭제했습니다.

honux77 commented on src/main/java/chess/board/Position.java:
```diff
@@ -0,0 +1,40 @@
+    }
+
+    public Position(String positionId) {
+        columnId = positionId.charAt(0);
```
this. 을 명시적으로 붙이는 습관이 가독성 측면에서 좋은 듯?

honux77 replied:
"A5" 이런 거 좋네요.

honux77 commented on src/main/java/chess/board/Position.java:
```diff
@@ -0,0 +1,40 @@
+
+public class Position {
+    private final char columnId;
+    private final int rowId;
```
역시 VO니까 생성자를 사용하는 것보다 create 메소드를 이용해서 만들어 주는 편이 좋을 것 같아요.

ghojeong replied:
VO 는 언제나 생성자가 아닌, 메소드를 통해 만들 수 있도록 주의하겠습니다.

honux77 commented on src/main/java/chess/pieces/Color.java:
```diff
@@ -1,6 +1,18 @@
+
+    public String generateRepresentation(String representation) {
+        return representationStrategy.apply(representation);
+    }
```
오오 재밌네요 👍  저는 getRepresentation() 메소드를 만들고 변환해서 리턴할 것 같아요.

honux77 commented on src/main/java/chess/pieces/EmptyPiece.java:
```diff
@@ -0,0 +1,20 @@
+    private static class SingletonHelper {
+        private static final EmptyPiece instance = new EmptyPiece();
+    }
+}
```
요것도 좋네요!

honux77 commented on src/main/java/chess/pieces/Pawn.java:
```diff
@@ -1,17 +1,16 @@
     public Pawn(Color color) {
-        this.color = color;
+        super(color);
     }
```
이 코드 지우면 문제가 되나요??

ghojeong replied:
네 문제가 됩니다. Pawn 의 생성자를 지우게 되면, `new Pawn()` 과 `new Pawn(color)` 2개를 동시에 같이 사용할 수 없게 됩니다.
개인적으로 Pawn 에 default 생성자는 불필요하다고 생각이 드는데,
`public Pawn()` 을 삭제할까요?
그리고 Color 를 생성자 대신에 static 팩토리 함수로 생성하는 것처럼 Pawn 도 생성자 없이 생성되도록 하는게 좋을 지 여쭈어봅니다.

honux77 replied:
ㅇㅇ 그것도 괜찮을 듯 해요.

honux77 commented on src/main/java/chess/utils/StringUtils.java:
```diff
@@ -0,0 +1,7 @@
+public class StringUtils {
+    public static final String NEWLINE = System.lineSeparator();
+
+    private StringUtils() {}
```
👍

jihye-woo replied:
NEWLINE 유틸 클래스 변수를 지정해놓으니까 문자열 관련 메소드들의 가독성이 더 좋아지는 것 같아요! 참고해야겠습니다👍

honux77 commented on src/main/java/chess/view/BoardView.java:
```diff
@@ -0,0 +1,54 @@
+    private String getRepresentation(List list) {
+        StringBuilder sb = new StringBuilder();
+        Iterator itr = list.iterator();
+        for (int i = 0; i < 8; i++) {
```
여기도 8이 하드코딩되어있군요.

honux77 commented on src/test/java/chess/board/BoardTest.java:
```diff
@@ -0,0 +1,29 @@
+        assertThat(board.getWhitePawnsResult())
+                .isEqualTo("pppppppp");
+        assertThat(board.getBlackPawnsResult())
+                .isEqualTo("PPPPPPPP");
```
아~ 이런 테스트가 있어서 public 메소드였던 거군요.

ghojeong replied:
사실 저런 방식의 테스트 좋아하지 않는데, Lucas 에 있어서 그대로 따래해보았습니다.
저 테스트 코드를 삭제 및 수정하고,, getWhitePawsResult 와 getBlackPawnsResult 퍼블릭 메소드가 필요없도록,
리팩토링 하겠습니다.

honux77 reviewed (changes requested):
테스트도 잘 짜고 고민 많이 한 흔적도 있고 좋네요.
코멘트 참고해서 조금만 더 깔끔해 지도록 개선해 보세요.
저랑은 코딩스타일이 좀 다르지만 그건 신경쓰지 않아도 될 듯해요.

Commit:
[refactor] minor change

- int i = 'a' -> char columnId = 'a'
- Position 은 VO 이므로 생성자가 아니라, static 함수를 통해 받을 수 있도록 함

Commit:
[refactor] Add PositionPool; Position 인스턴스를 풀에서 가져오도록 함

기존에는 매번 Position 객체를 새로 만들어서, 내부의 값을 비교했는데,
이제는 Pool 에서 동일한 객체를 가져오므로,
equals 와 hashCode 를 오버라이드 할 필요가 없게 되었다.
또한, 생성을 할 때 피드백에 따라 8부터 1로 감소시킬 필요가 없게 되었다.

Commit:
[refactor] Add BoardFactory; Board의 initilize 로직을 BoardFactory 로 옮김

BoardFactory 가 static 으로 떡칠된 유틸 클래스가 되어버려서,
마음에 들지가 않는다. BoardFactory 가 적절한 상태를 가질 수 있도록,
객체 설계를 더 계선할 필요가 있어 보인다.

Commit:
[feat] Add BadPositionException

체스판의 범위에서 벗어난 위치를 입력받으면,
BadPositionException 이 발생하도록 런타임 예외를 추가함

Commit:
[feat] 없는 커맨드에 대한 예외처리 추가

존재하지 않는 명령을 받으면 "잘못 입력하셨습니다" 라고 출력
출력 후 다시 명령을 받을 수 있도록 함

Commit:
[refactor] enum Command 를 디폴트 접근제어로 바꿈

Command 를 외부 패키지에서 사용하기를 원하지 않아서 public 을 지움

ghojeong commented:
존재하지 않는 커맨드에 대해서 "잘못 입력하셨습니다." 라고 출력후
다시 입력받을 수 있도록 했습니다.

<img width="170" alt="스크린샷 2021-02-19 오후 4 35 11" src="https://user-images.githubusercontent.com/38595650/108473789-d7409f80-72d1-11eb-9751-c5fdbebf7dec.png">

ghojeong commented:
피드백 주신 내용들을 대부분을 반영한것 같습니다.
다시 리뷰 요청드립니다.
@honux77

Commit:
[refactor] Piece 와 Pawn 을 VO 로 리팩토링

생성자를 private 으로 만들어서,
of static 함수를 통해서만 값을 가져오도록 함

ghojeong commented:
참고로 이건 제 개인 취향인데,
Factory 는 create
Value Object 는 of
Builder 는 build
라는 메소드 명을 즐겨 사용합니다.

honux77 reviewed (approved):
여러가지로 많이 고쳤군요! 좋습니다!
